libraries/Microsoft.Bot.Builder/AppInsightsLoggerMiddleware.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.Extensibility;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Builder
+{
+    /// <summary>
+    /// Middleware for logging incoming activitites into Application Insights.
+    /// In addition, registers a service that other Application Insights components can log
+    /// telemetry.
+    /// If this component is not registered, visibility within the Bot is not logged.
+    /// </summary>
+    public class AppInsightsLoggerMiddleware : IMiddleware
+    {
+        public const string AppInsightServiceKey = "AppInsightsContext";
```

Comment:
Since this is going to be used as a key into the `TurnContext::Services`, I would suggest something a little more explicit to avoid any possible collisions. Perhaps:

```CSharp
internal static readonly AppInsightsServiceKey = $"{nameof(AppInsightsLoggerMiddleware).AppInsightsContext";
```

Reply:
Good point. I missed that.

Reply:
Great idea